PEP 677: Callable Type Syntax

stroxler opened the pull request:
The draft of our Callable Type Syntax PEP is in a complete enough state that it is worth moving discussion to github so we have a history of major revisions going forward.

There are still many edits coming and we do not yet have a PEP number, hence keeping this as a draft PR.

Currently open problems (as of December 9) that I plan to get to soon:
- inline references
- add more to the motivation section
- add more to the section comparing to other languages, and possibly touch on possible lambda syntax

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,494 @@
+Key downsides that led us to reject the idea include the following:
+- A large majority of use cases only use positional-only arguments, and this syntax would be more verbose for that use case, both because of requiring argument names and an explicit ``/``, for example ``(int, /) -> bool`` where our proposal allows ``(int) -> bool``
+- The requirement for explicit ``/`` for positional-only arguments has a high risk of causing frequent bugs - which often wouldn’t be detected by unit tests - where library authors would accidentally use types with named arguments.
+- Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in PEP 612 would have made this difficult.
```
```suggestion
- Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in :pep:`612` would have made this difficult.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,494 @@
+
+There is no runtime implementation yet. At a high level we are committed to the following by backward compatibility:
+- We will need new object types for both the callable type and concatenation type, tentatively defined in C and exposed as ``types.CallableType`` and ``types.CallableConcatenateType`` in a manner similar to ``types.UnionType``.
+- The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime apis almost exactly
```
```suggestion
- The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime APIs almost exactly:
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,494 @@
+        # type: (int, str) -> bool
+        ...
+
+At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with PEP 604
```
```suggestion
At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with :pep:`604`.
```

AlexWaygood reviewed (commented):
Some thoughts

stroxler commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,494 @@
+
+The ``Callable`` type is widely used. For example, in typeshed [#typeshed-stats]_ it is the fifth most common complex type, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
+
+Most of the other commonly used types have gotten improved syntax either via PEP 604 or PEP 525.``Callable`` is used heavily enough to similarly justify a more usable syntax.
```
I don't mind this change, but I'm not sure it's necessary.

Other PEPs seem to be mixed in whether they just use PEP XXX versus :pep:`XXX` and it appears to format the same way on the website regardless.

AlexWaygood replied:
Okay fair! I was just applying standard rst rules from other parts of Python docs -- didn't realise PEPs worked differently 🙂

AlexWaygood replied:
With this specific instance, though, I do also feel like "have gotten improved syntax" is slightly inelegant writing 🙂

stroxler replied:
Oh oops, I missed that part of the line.

I'l apply the wording change, and wait for more guidance with respect to the PEP references.

Thanks!

stroxler commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,494 @@
+
+This PEP introduces a concise and friendly syntax for callable types, supporting the same functionality as ``typing.Callable`` but with an arrow syntax inspired by the syntax for typed function signatures. This allows types like ``Callable[[int, str], bool]`` to be written ``(int, str) -> bool``.
+
+The proposed syntax supports all the functionality provided by ``typing.Callable`` and ``typing.Concatenate`` and could eventually replace them entirely.
```
Yeah I think this needs work, thanks for pointing it out.

I don't think we have consensus about whether to discuss deprecation. I'd rather not bring it up at all because that's not conventional (PEP 604 aims to deprecate Union but does not say so).

stroxler replied:
For now I opted to go with "drop-in replacement" language, which I think makes the point we want without hinting at deprecation. We'll see what other reviewers think :)

AlexWaygood replied:
Nice, I think the revised wording is great!

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+
+There are four major problems with the existing ``Callable`` type:
+- it is verbose, particularly for more complex function signatures.
```
```suggestion
- It is verbose, particularly for more complex function signatures.
```

suutari commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+The ``Callable`` type is widely used. For example, in typeshed [#typeshed-stats]_ it is the fifth most common complex type, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
+
+Most of the other commonly used types have had their syntax improved via either PEP 604 or PEP 525. ``Callable`` is used heavily enough to similarly justify a more usable syntax.
```
Should the PEP-525 actually be PEP-585?

stroxler replied:
Oops, yes!

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+There are four major problems with the existing ``Callable`` type:
+- it is verbose, particularly for more complex function signatures.
+- it requires an explicit import, something we no longer require for most of the other
```
```suggestion
- It requires an explicit import, something we no longer require for most of the other
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- it is verbose, particularly for more complex function signatures.
+- it requires an explicit import, something we no longer require for most of the other
+  very common types after PEP 604 (``|`` for ``Union`` types) and PEP 585 (generic collections)
+- it does not visually represent the way function headers are written.
```
```suggestion
- It does not visually represent the way function headers are written.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- it requires an explicit import, something we no longer require for most of the other
+  very common types after PEP 604 (``|`` for ``Union`` types) and PEP 585 (generic collections)
+- it does not visually represent the way function headers are written.
+- it relies on two levels of nested square brackets. This can be quite hard to read,
```
```suggestion
- It relies on two levels of nested square brackets. This can be quite hard to read,
```

AlexWaygood reviewed (commented):
A little more sub-editing 🙂

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Rationale
+=========
+
+The ``Callable`` type is widely used. For example, in typeshed [#typeshed-stats]_ it is the fifth most common complex type, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
```
Would it be worth noting that this statistic is based on analysis done in October 2021? (Things might always change! The heavy presence of `Optional`, `Tuple`, `Union` and `List` in typeshed is due to a bug in mypy that means PEP 585 and PEP 604 can't yet be used in all situations. But a new release of mypy is just around the corner...)

stroxler replied:
Good point. I think using current stats is most fair since we're arguing by analogy here, but we should be specific about the time

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any future proposal using a bare ``**<some_type>`` to type ``kwargs``. This seems acceptable because:
+- If we ever do want such a syntax, it would be clearer and to require an argument name anyway so that the type looks more similar to a function signature. In other words, if we ever support typing ``kwargs`` in callable types, we would prefer ``(int, **kwargs: str)`` rather than ``(int, **str)``.
+- PEP 646 unpack syntax would rule out using ``*<some_type>`` for ``args``, and the ``kwargs`` case is similar enough that this rules out a bare ``**<some_type>`` anyway.
```
```suggestion
- PEP 646 unpacking syntax would rule out using ``*<some_type>`` for ``args``. The ``kwargs`` case is similar enough that this rules out a bare ``**<some_type>`` anyway.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- The group that debated these proposals was split down the middle about whether these changes are even desirable:
+  - On the one hand they make callable types more expressive, but on the other hand they could easily confuse users who haven’t read the full specification of callable type syntax.
+  - We believe the simpler syntax proposed in this PEP, which introduces no new semantics and closely mimics syntax in other popular languages like Kotlin, Scala, and TypesScript, are much less likely to confuse users.
+- We intend to implement the current proposal in a way that is forward-compatible with the more complicated extended syntax. So if the community decides after more experience and discussion that we want the additional features they should be straightforward to propose in the future.
```
```suggestion
- We intend to implement the current proposal in a way that is forward-compatible with the more complicated extended syntax. If the community decides after more experience and discussion that we want the additional features, they should be straightforward to propose in the future.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+As such, the only thing we need for backward compatibility is to ensure that types specified via the new syntax behave the same as equivalent ``typing.Callable`` and ``typing.Concatenate`` values they intend to replace.
+
+There’s no particular interaction between this proposal and ``from __future__ import annotations`` - just like any other type annotation it will be unparsed to a string at module import, and ``typing.get_type_hints`` should correctly evaluate the resulting strings in cases where that is possible.
```
```suggestion
There is no particular interaction between this proposal and ``from __future__ import annotations`` - just like any other type annotation it will be unparsed to a string at module import, and ``typing.get_type_hints`` should correctly evaluate the resulting strings in cases where that is possible.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+There is no runtime implementation yet. At a high level we are committed to the following by backward compatibility:
+- We will need new object types for both the callable type and concatenation type, tentatively defined in C and exposed as ``types.CallableType`` and ``types.CallableConcatenateType`` in a manner similar to ``types.UnionType``.
+- The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime APIs almost exactly:
+  - The ``__repr__`` methods will differ and display the new builtin syntax
+  - But the ``__args__`` and ``__parameters__`` fields must behave the same
+  - And the indexing operation - which returns a new type object with concrete types substituted for various entries in ``__parameters__``, must also be the same.
```
```suggestion
  - The ``__repr__`` methods will differ and display the new builtin syntax;
  - But the ``__args__`` and ``__parameters__`` fields must behave the same;
  - And the indexing operation - which returns a new type object with concrete types substituted for various entries in ``__parameters__``, must also be the same.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+The new runtime objects to which this syntax evaluates will remain backward-compatible with the ``typing.Callable`` and ``typing.Concatenate`` types they replace, other than details like ``__repr__`` where some behavior change makes sense.
+
+But we also believe that we should have a new runtime API with more structured data access, since:
+- Callable types have a more complicated shape than other generics, especially given the behavior when using ``...`` and ``typing.Concatenate``
```
```suggestion
- Callable types have a more complicated shape than other generics, especially given the behavior when using ``...`` and ``typing.Concatenate``.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+The current reference implementation has a fully-functional parser and all edge cases presented here have been tested.
+
+But there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.
```
```suggestion
However, there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.
```

stroxler replied:
I think I'll skip this one for now, I had it drilled into me in high school that "But" is okay and "However" is bad, can't overcome that training :)

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+But there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.
+
+This is not part of the specification per se but is an important detail to add  ress in our implementation. The solution will likely involve adding ``invalid_.*`` rules to ``python.gram`` and customizing error messages.
```
```suggestion
This is not part of the specification *per se* but is an important detail to address in our implementation. The solution will likely involve adding ``invalid_.*`` rules to ``python.gram`` and customizing error messages.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+        # type: (int, str) -> bool
+        ...
+
+At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with PEP 604
```
```suggestion
At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types. However, we have since begun to do so, e.g. with PEP 604.
```

AlexWaygood reviewed (commented):
A few more things I spotted! Some of these are more elegance-of-writing points 🙂

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+==========
+
+
+The ``Callable`` type, defined as part of PEP 484, is one of the most commonly used complex types in ``typing`` alongside ``Union`` and collection types like ``Dict`` and ``List``.
```
```suggestion
The ``Callable`` type, introduced in PEP 484, is one of the most commonly used complex types in the ``typing`` module alongside ``Union`` and collection types such as ``Dict`` and ``List``.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- It relies on two levels of nested square brackets. This can be quite hard to read,
+  especially when the function arguments themselves have square brackets.
+
+It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
```
```suggestion
It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. If libraries use less precise types, it reduces the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
```

AlexWaygood replied:
```suggestion
It is common for library authors to make use of untyped or partially typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
+
+With a succinct, easy-to-use syntax, developers may be less likely to reach for poorly-typed options. Callable types may also be beginner-friendly if we make them look more like function headers, and like the arrow type syntax used by several other popular languages.
```
```suggestion
With a succinct, easy-to-use syntax, developers may be less likely to reach for poorly typed options. Callable types may also be beginner-friendly if we make them look more like function headers, and like the arrow type syntax used by several other popular languages.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Rationale
+=========
+
+The ``Callable`` type is widely used. For example, as of October 2021 it is the fifth most common complex type in typeshed [#typeshed-stats]_, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
```
```suggestion
The ``Callable`` type is widely used. For example, as of October 2021 it was the fifth most common complex type in typeshed [#typeshed-stats]_, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+Most of the other commonly used types have had their syntax improved via either PEP 604 or PEP 585. ``Callable`` is used heavily enough to similarly justify a more usable syntax.
+
+We chose to support all the existing semantics of ``typing.Callable``, without adding support for new features. We took this decision after examining how frequently each feature might be used in existing typed and untyped open-source code. We determined that the vast majority of use cases are covered.
```
```suggestion
In this proposal, we chose to support all the existing semantics of ``typing.Callable``, without adding support for new features. We took this decision after examining how frequently each feature might be used in existing typed and untyped open-source code. We determined that the vast majority of use cases are covered.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Typing Behavior
+---------------
+
+Inside of type checkers, the new syntax should be treated with exactly the same semantics as ``typing.Callable``.
```
```suggestion
Type checkers should treat the new syntax with exactly the same semantics as ``typing.Callable``.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+Inside of type checkers, the new syntax should be treated with exactly the same semantics as ``typing.Callable``.
+
+So a type checker should treat the following pairs exactly the same::
```
```suggestion
As such, a type checker should treat the following pairs exactly the same::
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Grammar and Ast
+---------------
+
+The new syntax we’re proposing can be described by these AST changes ::
```
```suggestion
The proposed new syntax can be described by these AST changes ::
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- ParamSpec and Concatenate: The next most important feature is good support for PEP 612 ``ParamSpec`` and ``Concatenate`` types like ``(**P) -> bool`` and ``(int, **P) -> bool``. These are common primarily because of the heavy use of decorator patterns in python code.
+- TypeVarTuples: The next most important feature, assuming PEP 646 is accepted, is for unpacked types which are common because of cases where a wrapper passes along ``*args`` to some other function.
+
+Features that other, more complicated proposals would support account for fewer than 2% of the use cases we found. These are already expressible using `Callback Protocols <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_, and since they aren’t common we decided that it made more sense to move forward with a simpler syntax.
```
```suggestion
Features that other, more complicated proposals would support account for fewer than 2% of the use cases we found. These are already expressible using `Callback Protocols <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_, and since they are uncommon we decided that it made more sense to move forward with a simpler syntax.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+We decided against proposing it for the following reasons:
+- The implementation would have been more difficult, and usage stats demonstrate that fewer than 3% of use cases would benefit from any of the added features.
+- The group that debated these proposals was split down the middle about whether these changes are even desirable:
+  - On the one hand they make callable types more expressive. On the other hand, they could easily confuse users who haven’t read the full specification of callable type syntax.
```
```suggestion
  - On the one hand, they make callable types more expressive. On the other hand, they could easily confuse users who have not read the full specification of callable type syntax.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+- The implementation would have been more difficult, and usage stats demonstrate that fewer than 3% of use cases would benefit from any of the added features.
+- The group that debated these proposals was split down the middle about whether these changes are even desirable:
+  - On the one hand they make callable types more expressive. On the other hand, they could easily confuse users who haven’t read the full specification of callable type syntax.
+  - We believe the simpler syntax proposed in this PEP, which introduces no new semantics and closely mimics syntax in other popular languages like Kotlin, Scala, and TypesScript, are much less likely to confuse users.
```
```suggestion
  - We believe the simpler syntax proposed in this PEP, which introduces no new semantics and closely mimics syntax in other popular languages like Kotlin, Scala, and TypesScript, is much less likely to confuse users.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Key downsides that led us to reject the idea include the following:
+- A large majority of use cases only use positional-only arguments, and this syntax would be more verbose for that use case, both because of requiring argument names and an explicit ``/``, for example ``(int, /) -> bool`` where our proposal allows ``(int) -> bool``
+- The requirement for explicit ``/`` for positional-only arguments has a high risk of causing frequent bugs - which often wouldn’t be detected by unit tests - where library authors would accidentally use types with named arguments.
+- Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in PEP 612 would have made this difficult.
```
```suggestion
- Our analysis suggests that support for ``ParamSpec`` is key, but the scoping rules laid out in PEP 612 would have made this difficult.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+--------------------------
+
+An idea we looked at very early on was to allow using functions as types. This may be a great idea, but we consider less an alternative to better callable types than a major improvement in the usability of Callable Protocols:
+- Using functions as types wouldn’t give us a new way of describing function types as first class values. Instead, they would require a function definition statement that effectively defines a type alias (much as a Callable Protocol class statement does).
```
```suggestion
- Using functions as types would not give us a new way of describing function types as first class values. Instead, they would require a function definition statement that effectively defines a type alias (much as a Callable Protocol class statement does).
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+An idea we looked at very early on was to allow using functions as types. This may be a great idea, but we consider less an alternative to better callable types than a major improvement in the usability of Callable Protocols:
+- Using functions as types wouldn’t give us a new way of describing function types as first class values. Instead, they would require a function definition statement that effectively defines a type alias (much as a Callable Protocol class statement does).
+- Functions-as-types would support almost exactly the same features that Callable Protocols do today: named, optional, and variadic args as well as the ability to define overloads.
+So we think that is an idea for a related PEP, but not a direct substitute for improved Callable syntax.
```
```suggestion
Therefore, we think that is an idea for a related PEP, but not a direct substitute for improved Callable syntax.
```

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+Other Proposals Considered
+--------------------------
+
+An idea we looked at very early on was to allow using functions as types. This may be a great idea, but we consider less an alternative to better callable types than a major improvement in the usability of Callable Protocols:
```
It's a little unclear what this "using functions as types" proposal is. Could you add a link to somewhere where it's fleshed out a little more? 🙂

stroxler replied:
Yep! Unfortunately I don't think there's much to link to, everyone in the typing-sig discussion knows this from some meeings and emails but that's about it. I will see if there's a good email to link to, otherwise we may need to write a small doc on this.

AlexWaygood replied:
Clearly not "everyone", since I'm on typing-sig 😉

AlexWaygood replied:
(I think I joined about halfway through the discussion on improved callable syntax)

stroxler replied:
Makes sense. It's an idea Łukasz Langa raised at the 2021 PyCon typing summit.

The idea is that we could allow writing something like
```
def MyCallbackProtocolType(x: int, /, y: str = ...): ...
```
in place of
```
class MyCallbackProtocolType(Protocol):
    def __call__(self, x: int, /, y: str = ...): ...
```
and then use the function itself as a type representing any callable with a compatible call signature.

AlexWaygood replied:
Ooh, that would be nice. But yes -- a great idea for a different PEP!

stroxler replied:
I wrote up a doc on the idea because I don't want to inline too much about this proposal, since it mostly deserves a separate discussion not intermingled with comments about this PEP.
https://docs.google.com/document/d/1rv6CCDnmLIeDrYlXe-QcyT0xNPSYAuO1EBYjU3imU5s/edit

I linked to the doc, and also added a bit more detail directly in the PEP

AlexWaygood replied:
Ah, that's much better now! And yes, I agree that it's better to keep the details of that proposal out of the main body of this PEP :)

AlexWaygood replied:
A thought: perhaps this Google Doc could become an appendix to the PEP?

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,492 @@
+
+We decided against it because this is not visually as similar to existing function header syntax. Moreover, it is visually similar to lambdas, which bind names with no parentheses: ``lambda x, y: x == y``.
+
+Another idea was a new “special string” syntax an puting the type inside of it, for example ``t”(int, str) -> bool”``. We rejected this because it is not as readable, and it doesn’t seem in line with guidance from the Steering Council on ensuring that type expressions do not diverge from the rest of Python syntax. [#python-types-and-runtime-guidance]_
```
```suggestion
Another idea was a new “special string” syntax an puting the type inside of it, for example ``t”(int, str) -> bool”``. We rejected this because it is not as readable, and seems out of step with guidance from the Steering Council on ensuring that type expressions do not diverge from the rest of Python's syntax. [#python-types-and-runtime-guidance]_
```

AlexWaygood reviewed (commented):
I *think* this is my final pass 😄 it's a great PEP, I'm rooting for it!

AlexWaygood commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,782 @@
+    V = TypeVar("R")
+    P = ParamSpec("P")
+
+    def with_retries(f: (**P) -> R) -> (bool, **P) -> R:
```
I'm generally extremely supportive of this PEP, but I worry about this example. The number of arrows in one signature makes my eyes glaze over. The thing is, I'm 90% sure that this will be vastly improved by appropriate syntax highlighting in an IDE — but maybe that should be mentioned?

suutari replied:
Also simple whitespace trick might make it more readable:
```
def with_retries(
    f: (**P) -> R,
) -> (bool, **P) -> R:
```

suutari replied:
Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

AlexWaygood replied:
> Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

Not currently, I don't think, sadly — though there's a proposal to rework the way they're rendered that would give us syntax highlighting 🙏 (though I'm not sure we'd gain much from that with this PEP, anyway, since we're proposing new syntax?) https://discuss.python.org/t/request-for-comment-making-pep-rendering-self-contained/10774

stroxler replied:
It's probably not the best intro example but my creativity gave out and the easiest realistic use case I could think of was a decorator.

We'll see if @pradeep90 can come up with a better one. The problem is that we're supposed to be highlighting the kind of situation where people give up and use partially-typed functions, which means the example has to be a little uglier than the basic intro-to-callable-type.

suutari commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,845 @@
+
+    from typing import Any, Callable, Concatenate, ParamSpec, TypeVar
+
+    V = TypeVar("R")
```
```suggestion
    R = TypeVar("R")
```

suutari commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,845 @@
+
+    from typing import Any, ParamSpec, TypeVar
+
+    V = TypeVar("R")
```
```suggestion
    R = TypeVar("R")
```

stroxler commented:
@JelleZijlstra  Marking this as ready for review.

We likely still have some work to do on wording in the Motivation section, and probably more readability edits. But the proposal itself is pretty complete and we need to share this with python-dev to keep moving forward.

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+static analysis tooling, and code review.
+
+One of the types `defined by PEP 484
+<https://www.python.org/dev/peps/pep-0484/#callable>`_ in the is
```
```suggestion
<https://www.python.org/dev/peps/pep-0484/#callable>`_ is
```

JelleZijlstra replied:
This crept back in

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+``typing.Callable``, which describes a callable value such as a
+function or a method. It takes two parameters as inputs but with the
+first parameter being either a placeholder like ``...`` or a list of
+types. For example: - ``Callable[..., int]`` indicates a funciton with
```
Should this be an indented list?

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+    f6: (int, *Ts, str) -> bool
+    f6: Callable[[int, *Ts, str], bool]
+
+Grammar and Ast
```
```suggestion
Grammar and AST
```

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+~~~~~~~~~~~~~~~~
+
+
+``->`` binds less tightly than other operators, both inside types and in function signatures::
```
Does this mean that the two examples below are equivalent? Would be good to make that explicit.

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+
+    f6: (int, *Ts, str) -> bool
+    f6: Callable[[int, *Ts, str], bool]
+
```
Maybe add an example with no arguments: `() -> bool`. Below you say that `(,) -> bool` is an error but not what to write instead.

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+  (much as a Callable Protocol class statement does).
+- Functions-as-types would support almost exactly the same features
+  that Callable Protocols do today: named, optional, and variadic args
+  as well as the ability to define overloads.
```
It's also not obvious how to support ParamSpec with this proposal.

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+Introducing type-strings
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Another idea was a new “special string” syntax an puting the type
```
```suggestion
Another idea was adding a new “special string” syntax and putting the type
```

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+<https://mail.python.org/archives/list/python-dev@python.org/thread/VBHJOS3LOXGVU6I4FABM6DKHH65GGCUB>`_
+for feedback.
+
+The proposed AST and runtime behavior is explicitly designed to be
```
This paragraph feels out of place; there's already discussion of forward compatibility above.

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -0,0 +1,851 @@
+Kotlin
+~~~~~~
+
+Function types in `Kotlin <https://kotlinlang.org/docs/lambdas.html>`_ permits
```
```suggestion
Function types in `Kotlin <https://kotlinlang.org/docs/lambdas.html>`_ permit
```

JelleZijlstra reviewed (changes requested):
Thanks! I have a couple of small comments. I'll also let Guido review as the sponsor.

brettcannon commented:
Please update `CODEOWNERS` to add Guido for future reviews on this PEP.

gvanrossum commented:
Please fix the test failures (probably markup issues).

JelleZijlstra commented on pep-XXXX.rst:
```diff
@@ -0,0 +1,855 @@
+PEP: 9999
```
You got an extra stray file

stroxler replied:
Yep, just realized that and removed it. Thanks!

JelleZijlstra commented on pep-0677.rst:
```diff
@@ -38,14 +38,17 @@ bugs as well as helping with editor tooling like tab completion,
-``str`` and ``int`` and returning a ``bool``.
+types. For example:
+
+- ``Callable[..., int]`` indicates a funciton with arbitrary
```
```suggestion
- ``Callable[..., int]`` indicates a function with arbitrary
```

stroxler commented:
I think I fixed most of the build issues.

It's unhappy with the header, and I'm lost - as far as I can tell, I pasted the formatting exactly from an existing pep, and checked it against a few others. But it complains about indentation and block quotes on almost every line, starting with my name.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,855 @@
+
+We confirmed that the current proposal is forward-compatible with
+extended syntax by
+`implementing<https://github.com/stroxler/cpython/tree/callable-type-syntax--extended>`_
```
```suggestion
`implementing <https://github.com/stroxler/cpython/tree/callable-type-syntax--extended>`_
```

gvanrossum commented:
Can you address Jelle's change request?

stroxler commented:
I think I addressed the comments already.

I'm not sure how to tell github about it though (and I accidentally requested re-review while I was trying to figure it out, which I didn't intend to do).

The docs say there's a way to "dismiss review" once you've addressed comments but I cannot find it. Maybe I don't have the right permissions?

JelleZijlstra commented:
@gvanrossum do you want to do another round of review? If not I can merge the PR once CI is green and I can't find any more typos.

@stroxler I think you put the "in the" typo back at least once, let me check if it's still there.

gvanrossum commented:
IIRC you leave a comment saying
```
I have made the requested changes
```

stroxler commented:
I have made the requested changes

gvanrossum commented:
I dunno. I'll skim the thing one more time and then land it regardless.

JelleZijlstra commented:
(FYI, "I have made the requested changes" triggers the Bedevere bot on the CPython repo but it doesn't do anything on GiitHub in general.)

stroxler commented:
Awesome, thanks @JelleZijlstra and @gvanrossum!

If readability PRs later on are a problem then it may be worth waiting @pradeep90 has another go at the Motivation section, it's new and I think we'll want to reword it.

But if that's not a problem I'd prefer to land and send to python-dev, I think the sections they'll care about should be ready once you've reread them.

JelleZijlstra commented:
Later PRs are fine, I'm happy to review them. Let's just land this before somebody else grabs our number. :)

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+Grammar and AST
+---------------
+
+The proposed new syntax can be described by these AST changes ::
```
Might mention this goes into Parser/Python.asdl.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+``callable_type_positional_argument`` as follows::
+
+    callable_type_positional_argument:
+        | expression ','
+        | expression &')'
```
What happened to the `!'...'` guard?

stroxler replied:
Whoops - this section predates me realizing we needed that guard. Will fix!

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+to an operator like ``|``::
+
+    (int) -> bool | () -> bool    # syntax error!
+    (int) -> bool | (() -> bool)  # okay
```
So that's a function from (int) to a union type, the right half of which is another function.

Maybe a more useful example would be the union of two function types?

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+concerns, our grammar rules it out and the following is a syntax
+error::
+
+    (int, ...) -> bool
```
One more reason to disallow this is that it reminds of `tuple[int, ...]`.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+Incompatibility with other possible uses of ``*`` and ``**``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any
+future proposal using a bare ``**<some_type>`` to type
+``kwargs``. This seems acceptable because:
```
Hm. Occasionally we hear from someone who has a TypedDict that they want to use as the type for some function's kwargs. If we wanted to support that for a function *type*'s kwargs we'd be hosed -- the obvious notation would be `(**TD) -> bool`. Or maybe that would just work, and what you're ruling out here is use of `(**int) -> bool` to type e.g. `def f(**args: int) -> bool: ...`?

pradeep90 replied:
Hmm... Technically, there is nothing to prevent us from supporting `**KwargsTypedDict` (or even `**int`) in the future. Type checkers could distinguish `**KwargsTypedDict` and `**P` based on the actual values of the variables. 

The main reason we have `**P` instead of a bare `P` is to make it syntactically obvious that it is a ParamSpec (like `Callable[P, R]`). Otherwise, users might have trouble realizing that P is a ParamSpec in `(Foo, P) -> int`.

We could update the PEP (in a new PR) with the above.

stroxler replied:
I'm on the fence about this. Technically, there's nothing prevent us from using `**<some_type>` with divergent semantics depending on the information contained within `<some_type>`, type-checkers could handle this.

But it would be pretty unusual for python to let the semantics of the same syntax vary that much - a `ParamSpec` is a pretty different concept from typing the `kwarg` variadic argument.

My guess is that, in an extended syntax, we would work around this by requiring names for the vararg and kwarg,
so that we would use:
```
(**kwags: int) -> bool
(**kwargs: **MyTypedDict) -> bool
```

I think we should update the PEP to clarify that it's technically still possible to use `**int` and `**MyTypedDict` but I don't want to imply this is a good idea or say anything that's likely to prompt much discussion of it.

JelleZijlstra replied:
For what it's worth, I don't see the issue with allowing `**MyTypedDict` in the future: just like `**P`, it means "this is the type for all of the kwargs". Perhaps to keep this option open we should avoid using the name "paramspec" in the AST and runtime objects.

gvanrossum replied:
> But it would be pretty unusual for python to let the semantics of the same syntax vary that much

Would it, though? We have `1+2 == 3`, vs `"a" + "b" == "ab"`. We have `a[i]` (indexing list or tuple) vs. `d[key]` (lookup in dict or mapping) vs. `list[int]` (generic type). Etc. So I think there are enough precedents -- we don't need to worry about this looking odd.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+worth considering what possibilities would be ruled out by adopting
+this proposal.
+
+It would be incompatible with this proposal to adopt the same a
```
```suggestion
It would be incompatible with this proposal to adopt a
```

pradeep90 replied:
Can you explain why? (Because `(int) -> str` is a valid expression.)

-------

That reminds me: we need to have a subsection for aliases (`Foo = (int) -> str`, generic aliases) and how a callable type behaves at runtime when used as an expression. 

@gvanrossum  might have thoughts on things we should cover (especially from python-dev's point of view).

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+types, but Python is unique in that types are expressions and have to
+evaluate to runtime values. Our view is that this merits using
+separate tokens, and given the existing use of ``->`` for return types
+in function signatures it would be more coherent to use ``->`` for
```
```suggestion
in function signatures it would be more consistent to use ``->`` for
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+  where a wrapper passes along ``*args`` to some other function.
+
+Features that other, more complicated proposals would support account
+for fewer than 2% of the use cases we found. These are already
```
I wish we had more data to support the 2% number, since I still have a nagging doubt about it. I worry that this number will be quoted endlessly regardless of its truth value. Maybe Pradeep can publish a set of raw data sources one can use to repeat his analysis? (It would be cool if we could repeat the same analysis at some future point and see if the number is changed.)

pradeep90 replied:
[Guido]

I'd already shared the [script](https://github.com/pradeep90/annotation_collector#calls-to-callback-parameters) to repeat the analysis on any project from the command-line. The list of OSS projects with stats is [here](https://github.com/pradeep90/annotation_collector#stats) (raw [callables](https://github.com/pradeep90/annotation_collector/tree/master/data)). I definitely want people to repeat the analysis on their favorite code bases.

(@stroxler Could we explicitly link to the script in the resources and maybe drop a couple of lines describing the methods, the assortment of projects, and the limitations? We're referring to the stats quite a bit and people might be curious.)

[Guido]

> It would be cool if we could repeat the same analysis at some future point and see if the number is changed.

Yes, we can rerun the above script later on to see how usage has changed.

> I worry that this number will be quoted endlessly

+1. Feel like we should qualify it instead of presenting it as a bare stat, especially since people will question it. Maybe something like "X% of the cases we found in the N open-source projects we analyzed". That makes it explicit that YMMV and that this is an average.

> since I still have a nagging doubt about it

Were there particular types of projects where you feel the distribution would be different? It can help to test the hypothesis directly instead of letting the doubt nag :)

(For example, I had been curious how untyped projects used callbacks, which is why I included untyped projects like django. Same for open-source vs internal. Steven had been curious whether application code differed much from library code.)

gvanrossum replied:
I need to give up on fighting this on a numeric basis. I agree that it's probably little-used.

I still worry about confirmation bias, but my biggest, intuitive objection is that I feel the extended form is needed for *completeness* (i.e., there are function types whose values you can spell with a single `def` but whose type must be spelled with a callback protocol). I know it's fashionable to make fun of completionism, but I suspect that a lot of other useful features wouldn't have been in the language if we had based our decisions purely on data about current usage.

My second objection is that callback protocols are both ugly and unintuitive.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+We decided against proposing it for the following reasons:
+
+- The implementation would have been more difficult, and usage stats
```
```suggestion
- The implementation would have been more complex, and usage stats
```

pradeep90 replied:
```
- Complexity: The implementation ...
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+We decided against proposing it for the following reasons:
+
+- The implementation would have been more difficult, and usage stats
+  demonstrate that fewer than 3% of use cases would benefit from any
```
Same feeling about the 3% number.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+    less likely to confuse users.
+
+- We intend to implement the current proposal in a way that is
+  forward-compatible with the more complicated extended syntax. If the
```
```suggestion
  forward-compatible with the more complex extended syntax. If the
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+  community decides after more experience and discussion that we want
+  the additional features, they should be straightforward to propose
+  in the future.
+- We realized that because of overloads, it is not possible to replace
```
I wonder if "because of overloads" requires a complete example (e.g. an overloaded function, the observation that its type can't be expressed using current function types, and the conclusion that it must be done using a Callback Protocol). It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

pradeep90 replied:
+1 on a concrete example.

> It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

Overloads correspond more to an intersection of callable types. If we had a union `f: (int) -> str | ((list[int]) -> None)`, then we wouldn't be able to pass an argument to just one of the union items: `f(42)` would error because it's not a `list[int]`; `f([1, 2])` would error because it's not an `int`.

pradeep90 replied:
```
- No replacement for callback protocol:
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+We considered adding support for named, optional, and variadic
+arguments. However, we decided against including these features, as
+our analysis showed they are infrequently used. When they are really
+needed, it is possible to type these using `Callback Protocols
+<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
```
I think neither Callback nor Protocols should be capitalized. This would have to change throughout the PEP.

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+Functions-as-Types
+~~~~~~~~~~~~~~~~~~
+
+An idea we looked at very early on was to `allow using functions as
+types
```
I'd combine these two lines.
```suggestion
An idea we looked at very early on was to `allow using functions as types
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+    (Int, String) => Bool
+
+Scala, like Python, has the ability to provide function arguments by name.
+Funciton types can optionally include names, for example::
```
```suggestion
Function types can optionally include names, for example::
```

gvanrossum commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+<https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/fsharp-types#syntax-for-types>`_,
+`OCaml
+<https://www2.ocaml.org/learn/tutorials/basics.html#Defining-a-function>`_,
+and ``Haskell <https://wiki.haskell.org/Type_signature>`_, all use
```
```suggestion
and `Haskell <https://wiki.haskell.org/Type_signature>`_, all use
```
(This markup bug looks serious, and I'm going to commit it.)

gvanrossum reviewed (approved):
Despite various comments I'm going to land this, after fixing only one definite markup bug.

gvanrossum commented:
Congrats! I wish all PEP authors were like you. :-)

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+
+``->`` binds less tightly than other operators, both inside types and
+in function signatures, so the following two callable types are
+equivalent::
```
Can we explain it in words? This section is pretty dense and people may not know what the result is after it "binds tightly".

"..., so `(int) -> str | bool` refers to a callable that accepts `int` and returns `str | bool` rather than a union of `(int) -> str` and `bool`. That is::

    (int) -> str | bool

    # equivalent to
    (int) -> (str | bool)

    # NOT
     ((int) -> str) | bool

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+    (int) -> (str | bool)
+
+
+``->`` associates to the right, both inside types and in function
+signatures. So the following pairs are equivalent::
```
Same. Would appreciate mentioning it in words before introducing "associates to the right" (since that's usually in the context of arithmetic-like operators):

"``->`` associates to the right, both inside .... So, `(int) -> (str) -> bool` is a callable that accepts `int` and returns another callable `(str) -> bool`. The following pairs are equivalent::

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+``async`` Keyword
+~~~~~~~~~~~~~~~~~
+
+All of the binding rules still work for async callable types::
```
... The following pairs are equivalent::

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+    from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple
+
+    P = ParamSpec("P")
```
Feel like it is worth adding one line before each pair explaining what it is that, say, `() -> bool` means.

Five years from now, if all goes to plan, nobody will even know what `typing.Callable[[], bool]` meant 🤣  

So, it's our job to explain what it means. For complex stuff like ParamSpec, we can at least link to the PEP for more details.

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+Trailing Commas
+~~~~~~~~~~~~~~~
+
+- Following the precedent of function signatures, putting a comma in
+  an empty arguments list is illegal, ``(,) -> bool`` is a syntax
+  error.
+- Again following precedent, trailing commas are otherwise always
+  permitted::
```
"precedent" is an unfortunate choice of word here :) Maybe just "the example of function signatures"?

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+  permitted::
+
+
+    ((int,) -> bool == (int) -> bool
+    ((int, **P,) -> bool == (int, **P) -> bool
+    ((...,) -> bool) == ((...) -> bool)
+
```
Can we stick to the pair of equivalent lines formula?

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under normal circumstances, any valid expression is permitted where we
+want a type annotation and ``...`` is a valid expression. This is
```
^ For example, "(int, ...) -> bool".

Otherwise, the sentence is quite abstract.

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+never semantically valid and all type checkers would reject it, but
+the grammar would allow it if we did not explicitly prevent this.
+
+We decided that there were compelling reasons to prevent it: - The
```
Formatting: needs to be a list.

PS: You can run `make rss` locally to run the CI checks (even though it wouldn't have caught this).

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+expressible using Callback Protocols, and since they are uncommon we
+decided that it made more sense to move forward with a simpler syntax.
+
+Extended Syntax Supporting Named and Optional Arguments
```
nit: lower case "Supporting"

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+types would have been equivalent::
+
+    class Function(typing.Protocol):
+        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
```
`s/f/__call__/`

`s/Function/MyCallback`

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
+            ...
+
+    Function = (int, y: float, *, z: bool = ..., **kwargs: str) -> bool
```
s/Function/MyCallback/

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+    Function = (int, y: float, *, z: bool = ..., **kwargs: str) -> bool
+
+Advantages of this syntax include: - Most of the advantages of the
```
Formatting: Needs to be a list.

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+- We intend to implement the current proposal in a way that is
+  forward-compatible with the more complicated extended syntax. If the
+  community decides after more experience and discussion that we want
+  the additional features, they should be straightforward to propose
```
Ambiguous pronoun "they" (I thought it was "community").

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+In this proposal, the following types would have been equivalent::
+
+    class Function(typing.Protocol):
+        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
+            ...
+
+    Function = (x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool
```
Same substitutions in the previous example: `s/Function/MyCallback/g` and `s/f/__call__`

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+- The implementation would have been more difficult, and usage stats
+  demonstrate that fewer than 3% of use cases would benefit from any
+  of the added features.
+- The group that debated these proposals was split down the middle
```
```
- Lack of consensus: The group that debated these proposals (typing-sig) was split ...
```

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+    popular languages like Kotlin, Scala, and TypesScript, is much
+    less likely to confuse users.
+
+- We intend to implement the current proposal in a way that is
```
```
- Forward-compatibility:
```

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+------------------------------------
+
+One alternative we had floated was a syntax much more similar to
+function signatures.
+
+In this proposal, the following types would have been equivalent::
```
Merge into a single paragraph?

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+An idea we looked at very early on was to `allow using functions as
+types
+<https://docs.google.com/document/d/1rv6CCDnmLIeDrYlXe-QcyT0xNPSYAuO1EBYjU3imU5s/edit?usp=sharing>`_.
+The idea is allowing a function to stand in for its own call
+signature, with roughly the same semantics as the ``__call__`` method
+of Callback Protocols. Think this may be a great idea and worth its
+own PEP, but that it is not a good alternative to improving the
+usability of callable types:
```
Quick example might be good (just like the other callback protocol snippets).

Can you make this sentence less informal? "Think this may be a great idea ..."

stroxler replied:
Yeah, I shied away from an example because it looks kinda weird if we snake case (which is always pythonic for functions, never pythonic for types), and it also looks kinda weird if we camel case (which is always pythonic for types, never pythonic for functions). But I guess I should pick one and just look weird.

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+usability of callable types:
+
+- Using functions as types would not give us a new way of describing
+  function types as first class values. Instead, they would require a
```
... first-class values. That is, we would not be able to define them anywhere within a type. Instead, they would require a separate, out-of-line ...

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+  function types as first class values. Instead, they would require a
+  function definition statement that effectively defines a type alias
+  (much as a Callable Protocol class statement does).
+- Functions-as-types would support almost exactly the same features
```
This right now reads positively. I was left a bit confused since this looked like a list of "Cons".

Maybe rephrase as "They would not support any more features than callback protocols while only being somewhat shorter."

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+  that Callable Protocols do today: named, optional, and variadic args
+  as well as the ability to define overloads.
+
+Another reason we don't view functions-as-types as a good alternative
```
Bullet point?

pradeep90 commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+**Pradeep** `brought this proposal to python-dev
+<https://mail.python.org/archives/list/python-dev@python.org/thread/VBHJOS3LOXGVU6I4FABM6DKHH65GGCUB>`_
+for feedback.
+
```
A Usage Statistics section would be good to describe the methods, projects, and [limitations](https://github.com/pradeep90/annotation_collector#limitations). We could link to it from other parts of the PEP.

stroxler commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
+
+See the Rejected Alternatives section for more detailed discussion
+about omitted features.
```
Done

carljm commented on pep-0677.rst:
```diff
@@ -0,0 +1,854 @@
+
+    from app_logic import Response, UserSetting
+
+    def make_endpoint(
```
Not sure why this comment is marked resolved, we still have `async def customize_response_for_settings` above and `def make_endpoint` here. AFAICS the difference is irrelevant to the example and both should be just `def make_endpoint`

pradeep90 commented:
@Carl Meyer ***@***.***> Steven marked it resolved because he fixed
it in the follow-up PR: https://github.com/python/peps/pull/2193. Sorry for
the confusion; I reviewed this PR late.

On Thu, Dec 16, 2021 at 10:03 AM Carl Meyer ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pep-0677.rst
> <https://github.com/python/peps/pull/2187#discussion_r770788940>:
>
> > +
> +    from typing import Awaitable, Callable
> +    from app_logic import Response, UserSetting
> +
> +
> +    async def customize_response_for_settings(
> +        response: Response,
> +        customizer: Callable[[Response, list[UserSetting]], Awaitable[Response]]
> +    ) -> Response:
> +       ...
> +
> +With our proposal, this code can be abbreviated to::
> +
> +    from app_logic import Response, UserSetting
> +
> +    def make_endpoint(
>
> Not sure why this comment is marked resolved, we still have async def
> customize_response_for_settings above and def make_endpoint here. AFAICS
> the difference is irrelevant to the example and both should be just def
> make_endpoint
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/python/peps/pull/2187#discussion_r770788940>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC6L67QTZZ4FYTNLZA2XJTURISY3ANCNFSM5JVN4N7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
S Pradeep Kumar
